File: docs/aggregate_reports.mjs
// Generate one MyST page per funded project by concatenating its reports newest-first.
//
// Each project's reports are listed in its project.yml as a glob or a list of
// filenames/globs (reports: "*-report.md", or reports: [2026-06-report.md, ...]).
// Each report carries its own date in its frontmatter (date: "2026-06", or
// "2026-06-03" when the day matters).
//
// Reads ../Reports/<cycle>/<project>/ and writes ./projects/<cycle>/<project>.md.

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { mystParse } from "myst-parser";
import { parse as parseYaml } from "yaml";

const DOCS = path.dirname(fileURLToPath(import.meta.url));
const REPORTS = path.join(DOCS, "..", "Reports");
const OUT = path.join(DOCS, "projects");

const MONTHS = ["January", "February", "March", "April", "May", "June",
                "July", "August", "September", "October", "November", "December"];

// Read project.yml (title, proposal_issue, status, reports), falling back to the folder name.
function projectConfig(project) {
  const config = path.join(project, "project.yml");
  const info = fs.existsSync(config) ? parseYaml(fs.readFileSync(config, "utf8")) : {};
  if (!info.title) {
    console.log(`!! no title in ${config}; using folder name`);
    info.title = path.basename(project).replaceAll("_", " ").replace(/\b\w/g, (c) => c.toUpperCase());
  }
  return info;
}

// Pull the report's date out of its frontmatter: date: "2026-06" or "2026-06-03".
function reportDate(text) {
  if (!text.startsWith("---\n")) return null;
  const end = text.indexOf("\n---\n", 4);
  if (end === -1) return null;
  return text.slice(4, end).match(/^date:\s*["']?(\d{4})-(\d{2})(?:-(\d{2}))?["']?\s*$/m);
}

function stripFrontmatter(text) {
  if (text.startsWith("---\n")) {
    const end = text.indexOf("\n---\n", 4);
    if (end !== -1) return text.slice(end + 5).replace(/^\n+/, "");
  }
  return text;
}

// Shift headings down one level so report sections sit under the month heading.
function demoteHeadings(text) {
  const lines = text.split("\n");
  // We only need to find top-level headings (though we unwrap top-level +++ blocks as well).
  const top = mystParse(text).children.flatMap((n) => (n.type === "block" ? n.children : [n]));
  for (const node of top) {
    if (node.type === "heading" && node.depth < 6) {
      // Doubling the line's first "#" demotes an ATX heading. A setext
      // (underlined) heading has no "#", so it keeps its level.
      const start = node.position.start.line - 1;
      lines[start] = lines[start].replace("#", "##");
    }
  }
  return lines.join("\n");
}

fs.rmSync(OUT, { recursive: true, force: true });
const dirNames = (dir) =>
  fs.readdirSync(dir, { withFileTypes: true }).filter((d) => d.isDirectory()).map((d) => d.name).sort();

for (const cycle of dirNames(REPORTS)) {
  for (const projectName of dirNames(path.join(REPORTS, cycle))) {
    const project = path.join(REPORTS, cycle, projectName);
    const info = projectConfig(project);
    const files = new Set();
    for (const pattern of [info.reports ?? []].flat()) {
      const matches = fs.globSync(pattern, { cwd: project }).sort();
      if (!matches.length) console.log(`!! no report matches "${pattern}" in ${project}`);
      for (const f of matches) files.add(f);
    }
    const reports = [];
    for (const name of files) {
      const file = path.join(project, name);
      const text = fs.readFileSync(file, "utf8");
      const m = reportDate(text);
      if (!m) {
        console.log(`!! no date in the frontmatter of ${file}`);
        continue;
      }
      reports.push({ year: m[1], month: m[2], day: m[3], text });
    }
    // Catch report files that were added to the folder but never listed/matched.
    for (const md of fs.readdirSync(project).filter((f) => f.endsWith(".md") && !files.has(f))) {
      console.log(`!! ${path.join(project, md)} is not matched by reports in project.yml`);
    }
    if (!reports.length) continue;
    reports.sort((a, b) =>
      `${b.year}-${b.month}-${b.day ?? ""}`.localeCompare(`${a.year}-${a.month}-${a.day ?? ""}`));
    const sections = reports.map((r) => {
      const body = demoteHeadings(stripFrontmatter(r.text));
      return `## ${MONTHS[Number(r.month) - 1]} ${r.year}\n\n${body.trim()}\n`;
    });
    // Only claim the precision the report's date gives: "July 2, 2026" or
    // "June 2026". The ISO "latest report date" field exists solely for
    // the index listing's :sort:; it is never displayed.
    const { year, month, day } = reports[0];
    const monthName = MONTHS[Number(month) - 1];
    const latest = day ? `${monthName} ${Number(day)}, ${year}` : `${monthName} ${year}`;
    const latestIso = day ? `${year}-${month}-${day}` : `${year}-${month}`;
    // Frontmatter field names double as the index listing's column headers.
    const round = info.funding_round ?? "";
    if (!round) console.log(`!! no funding_round in ${path.join(project, "project.yml")}`);
    const frontmatter = [
      `title: "${info.title}"`,
      `funding round: "${round}"`,
      `latest report: "${latest}"`,
      `latest report date: "${latestIso}"`,
    ];
    const intro = [];
    if (info.status) {
      frontmatter.push(`status: "${info.status}"`);
      intro.push(`**Status:** ${info.status}`);
    }
    if (info.proposal_issue) intro.push(`[Proposal](${info.proposal_issue})`);
    let header = `---\n${frontmatter.join("\n")}\n---\n\n`;
    if (intro.length) header += `${intro.join(" · ")}\n\n`;
    const page = path.join(OUT, cycle, `${projectName}.md`);
    fs.mkdirSync(path.dirname(page), { recursive: true });
    fs.writeFileSync(page, header + sections.join("\n"));
    console.log(`wrote ${path.relative(DOCS, page)} (${reports.length} reports)`);
  }
}

File: docs/aggregate_reports.py
"""Generate one MyST page per funded project by concatenating its reports in date order.

Reads Reports/<cycle>/<project>/*.md and writes docs/projects/<cycle>/<project>.md.
Run from anywhere: paths are resolved relative to this file.
"""

import re
import shutil
import tomllib
from pathlib import Path

DOCS = Path(__file__).resolve().parent
REPORTS = DOCS.parent / "Reports"
OUT = DOCS / "projects"

MONTHS = ["January", "February", "March", "April", "May", "June",
          "July", "August", "September", "October", "November", "December"]


def project_config(project):
    """Read project.toml (title, proposal, status), falling back to the folder name."""
    config = project / "project.toml"
    info = tomllib.loads(config.read_text()) if config.is_file() else {}
    if not info.get("title"):
        print(f"!! no title in {config}; using folder name")
        info["title"] = project.name.replace("_", " ").title()
    return info


def strip_frontmatter(text):
    if text.startswith("---\n"):
        end = text.find("\n---\n", 4)
        if end != -1:
            return text[end + 5:].lstrip("\n")
    return text


def demote_headings(text):
    """Shift headings down one level so report sections sit under the month heading.

    Skips fenced code blocks so e.g. shell comments aren't rewritten.
    """
    lines = text.splitlines(keepends=True)
    fence = None
    for i, line in enumerate(lines):
        stripped = line.lstrip()
        if fence:
            if stripped.startswith(fence):
                fence = None
        elif stripped.startswith(("```", "~~~")):
            fence = stripped[:3]
        else:
            lines[i] = re.sub(r"^(#{1,5}) ", r"#\1 ", line)
    return "".join(lines)


def main():
    shutil.rmtree(OUT, ignore_errors=True)
    for cycle in sorted(REPORTS.iterdir()):
        if not cycle.is_dir():
            continue
        for project in sorted(cycle.iterdir()):
            if not project.is_dir():
                continue
            info = project_config(project)
            name = info["title"]
            reports = []
            for md in project.glob("*.md"):
                m = re.match(r"(\d{4})-(\d{2})(?:-(\d{2}))?", md.name)
                if not m:
                    print(f"!! skipping unrecognized filename: {md}")
                    continue
                reports.append(((m.group(1), m.group(2), m.group(3)), md))
            if not reports:
                continue
            reports.sort(key=lambda r: (r[0][0], r[0][1], r[0][2] or ""))
            sections = []
            for (year, month, _), md in reports:
                body = demote_headings(strip_frontmatter(md.read_text()))
                sections.append(f"## {MONTHS[int(month) - 1]} {year}\n\n{body.strip()}\n")
            # Only claim the precision the filename gives: "July 2, 2026" or
            # "June 2026". The ISO "latest report date" field exists solely for
            # the index listing's :sort:; it is never displayed.
            year, month, day = reports[-1][0]
            month_name = MONTHS[int(month) - 1]
            latest = f"{month_name} {int(day)}, {year}" if day else f"{month_name} {year}"
            latest_iso = f"{year}-{month}-{day}" if day else f"{year}-{month}"
            # Frontmatter field names double as the index listing's column
            # headers.
            round_ = info.get("funding_round", "")
            if not round_:
                print(f"!! no funding_round in {project / 'project.toml'}")
            frontmatter = [
                f'title: "{name}"',
                f'funding round: "{round_}"',
                f'latest report: "{latest}"',
                f'latest report date: "{latest_iso}"',
            ]
            intro = []
            if info.get("status"):
                frontmatter.append(f'status: "{info["status"]}"')
                intro.append(f'**Status:** {info["status"]}')
            if info.get("proposal"):
                intro.append(f'[Proposal]({info["proposal"]})')
            header = "---\n" + "\n".join(frontmatter) + "\n---\n\n"
            if intro:
                header += " · ".join(intro) + "\n\n"
            page = OUT / cycle.name / f"{project.name}.md"
            page.parent.mkdir(parents=True, exist_ok=True)
            page.write_text(header + "\n".join(sections))
            print(f"wrote {page.relative_to(DOCS)} ({len(reports)} reports)")


if __name__ == "__main__":
    main()
